### PR TITLE
Some large changes to the samplessheet test script to encorporate tso500 workflow

### PR DESCRIPTION
**Pull request by alexiswl**
Added function based hierarchy to script, added logging, pandas now runs a groupby over the non-samplesheet-column 'Type' to write each file to a csv file.
Header data is now read in as a list of rows and trailing commas are stripped.
Sheet names for the tracking df are no longer hardcoded - but now reads in any sheet that's a number that ISN'T 2018.
Pivot row in sample sheet now calculated by searching for [Data] or [Data],,,

### CODE REVIEW

**Comment by alexiswl** [comment]
@skanwal For some reason I cannot set assignees. Would you mind having a look over this? I've made some rather large changes to the script.

**Comment by skanwal** [comment]
@alexiswl - thanks.
I have assigned it to myself and will review it later today.

File: scripts/showcase/samplesheet-prep-test.py [hunk 1]
```diff
@@ -1,55 +1,399 @@
...
+    """
+    Read args
+    return: argparse dict
+    """
+    # Initialise parser
+    parser = argparse.ArgumentParser()
+    # Place arguments
+    parser.add_argument("--samplesheet", "-s", type=str, required=True)
+    parser.add_argument("--trackingsheet", "-t", type=str, required=True)
+    parser.add_argument("--outputFile", "-o", type=str, required=True)
```
Comment (skanwal) [inline scripts/showcase/samplesheet-prep-test.py]: update argument change to outputpath - as we are now writing multiple samplesheets

File: scripts/showcase/samplesheet-prep-test.py [hunk 2]
```diff
@@ -1,55 +1,399 @@
...
+    :param sample_sheet: path to the samplesheet
+    :return: An integer specifying which row the [Data] header is on.
+    """
+
+    backslash_chars = '[]'
+    escaped_header_line = ''.join(["\\%s" % char if char in backslash_chars else char
+                                   for char in HEADER_LINE_PRECURSOR
+                                  ])
+
+    re_pattern = re.compile("^%s(,)*" % escaped_header_line)
```
Comment (skanwal) [inline scripts/showcase/samplesheet-prep-test.py]: We can possibly get rid of this part, as discussed

File: scripts/showcase/samplesheet-prep-test.py [hunk 3]
```diff
@@ -1,55 +1,399 @@
...
+    """
+
+    with open(sample_sheet_output_file, 'w') as sample_sheet_output_h:
+        # Write out the header rows
+        sample_sheet_output_h.writelines(sample_sheet_header_rows)
+        # Write out [Data]
+        sample_sheet_output_h.write("{}\n".format(HEADER_LINE_PRECURSOR))
+        # Write out sample sheet
+        sample_sheet_df.\
+            to_csv(sample_sheet_output_h, sep=",", header=True, index=False)
```
Comment (skanwal) [inline scripts/showcase/samplesheet-prep-test.py]: Please merge this with the `write_sample_sheets` function

File: scripts/showcase/samplesheet-prep-test.py [hunk 4]
```diff
@@ -1,55 +1,399 @@
...
+    if modification_strategy == "TSO500":
+        # Add two extra columns to sample-sheet
+        modified_sample_sheet_df = convert_sample_sheet_to_tso500(sample_sheet_df)
+        # No changes to header rows required.
+        modify_sample_header_rows = sample_sheet_header_rows
+    elif modification_strategy == "V2":
+        # Add two extra columns to sample-sheet
+        modified_sample_sheet_df = convert_sample_sheet_to_v2(sample_sheet_df)
+        # No changes to header rows required.
+        modify_sample_header_rows = convert_sample_sheet_header_to_v2(sample_sheet_header_rows)
```
Comment (skanwal) [inline scripts/showcase/samplesheet-prep-test.py]: needs update to comments.

File: scripts/showcase/samplesheet-prep-test.py (hunk 1 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R72-R74

File: scripts/showcase/samplesheet-prep-test.py (hunk 2 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R72-R74

Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R122-R129

File: scripts/showcase/samplesheet-prep-test.py (hunk 3 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R354-R378

File: scripts/showcase/samplesheet-prep-test.py (hunk 4 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: https://github.com/umccr/infrastructure/pull/16/files#diff-9803ecbbdf15c05af6e26ccc75f88c84R342-R345

File: scripts/showcase/samplesheet-prep-test.py (hunk 1 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: `--outputFile` now changed to `--outputPath`.

File: scripts/showcase/samplesheet-prep-test.py (hunk 2 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: Yep we're using startswith("[Data]") now.

File: scripts/showcase/samplesheet-prep-test.py (hunk 3 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: Done! Preferably we'd have separate functions for the modification process and the write out process. There's a bit going on in those four lines at the bottom so having its own function with a doc string could make the code easier to read. Maybe a change in name from `write_sample_sheets` to `output_sample_sheets` and `write_sample_sheet` to `write_sample_sheet_to_file` would have reduced ambiguity around the function naming.

File: scripts/showcase/samplesheet-prep-test.py (hunk 4 above)
Comment (alexiswl) [reply scripts/showcase/samplesheet-prep-test.py]: Done, I've changed this around a little so now modification 'strategies' aren't necessarily mutually exclusive. **modification_strategy** isn't a great variable name in this case, happy to change it to something that's more descriptive of the different keys in the VALID_SAMPLE_TYPES global variable.

**Comment by alexiswl** [comment]
I've created a root task (samplesheet-splitter, tdn.aed4bd8ae94241ebb2c358bae95439b9) with two versions, v1 and v1-dev. The only difference being the input url for /mnt/samplehsheet-prep.py.

The following input.json resulted in a completed workflow with expected outputs
(trn.6d601c03abdb4d22b180ca365f5f15d0):
```
{
    "name": "taskRunName",
    "description": "This is a task run test of samplesheet-splitter, v1 (tdn.aed4bd8ae94241ebb2c358bae95439b9, tvn.43a9b2befbee4b0c9fc43a9b3d10cc1a)",
    "arguments": {
        "imageName":"umccr/alpine_pandas",
        "imageTag": "latest-slim",
        "runId": "191220_A00130_0127_BHMCNVDSXX",
        "gdsLogFolder": "gds://lucattini-dev/samplesheet-splitter/logs"
    }
}
```

**Comment by skanwal** [comment]
@alexiswl - have you posted this task to stratus `TES` repo? or somewhere else? I would like to give it a go. The changes look good bdw. TA.

**Comment by alexiswl** [comment]
> @alexiswl - have you posted this task to stratus `TES` repo? or somewhere else? I would like to give it a go. The changes look good bdw. TA.

It's in stratus as its own task, use v1-dev version for now as this points to the python script in my branch.

See the comment above for the inputs requirements:
> I've created a root task (samplesheet-splitter, tdn.aed4bd8ae94241ebb2c358bae95439b9) with two versions, v1 and v1-dev. The only difference being the input url for /mnt/samplehsheet-prep.py.
> 
> The following input.json resulted in a completed workflow with expected outputs
> (trn.6d601c03abdb4d22b180ca365f5f15d0):
> 
> ```
> {
>     "name": "taskRunName",
>     "description": "This is a task run test of samplesheet-splitter, v1-dev (tdn.aed4bd8ae94241ebb2c358bae95439b9, tvn.5528f47dc2084beea5ba19776029166b)",
>     "arguments": {
>         "imageName":"umccr/alpine_pandas",
>         "imageTag": "latest-slim",
>         "runId": "191220_A00130_0127_BHMCNVDSXX",
>         "gdsLogFolder": "gds://lucattini-dev/samplesheet-splitter/logs"
>     }
> }
> ```

Listed below is the root task ID (tdn) and the task versions (tvn). 

```
$ iap tasks list
...
id                                   | name 
tdn.aed4bd8ae94241ebb2c358bae95439b9 │ samplesheet-splitter    
```

```
$ iap tasks versions list tdn.aed4bd8ae94241ebb2c358bae95439b9
id                                   │ version     │ description
tvn.43a9b2befbee4b0c9fc43a9b3d10cc1a │ v1          │ Split the sample sheet
tvn.5528f47dc2084beea5ba19776029166b | v1-dev      | Split the sample sheet
```

**Pull request merged**